### PR TITLE
implement transaction message escape

### PR DESCRIPTION
**Pull request by azhsmesos**
docs: https://shimo.im/docs/rp3OVGvVLpFav8Am
具体开发思路在 proposal 下

### CODE REVIEW

File: broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionEscapeBridgeTest.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.apache.rocketmq.broker.transaction.queue;
+
+import org.junit.Ignore;
+
+/**
+ * @author zhaozhenhang <zhaozhenhang@kuaishou.com>
+ * Created on 2022-06-18
+ */
```
Comment (hzh0425) [inline broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionEscapeBridgeTest.java:8]: Don't leave author info

File: broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionEscapeBridgeTest.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package org.apache.rocketmq.broker.transaction.queue;
```
Comment (hzh0425) [inline broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionEscapeBridgeTest.java:1]: need apache Licence if add new class

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java [hunk 3]
```diff
@@ -354,7 +358,13 @@ private PutMessageResult putBackToHalfQueueReturnResult(MessageExt messageExt) {
...
         try {
             MessageExtBrokerInner msgInner = transactionalMessageBridge.renewHalfMessageInner(messageExt);
-            putMessageResult = transactionalMessageBridge.putMessageReturnResult(msgInner);
+            if (this.transactionalMessageBridge.getBrokerController().isSpecialServiceRunning()
+                    && BrokerRole.SLAVE == this.transactionalMessageBridge.getBrokerController().getMessageStoreConfig()
+                    .getBrokerRole()) {
+                putMessageResult = transactionalMessageBridge.getBrokerController().getEscapeBridge().putMessage(msgInner);
+            } else {
+                putMessageResult = transactionalMessageBridge.putMessageReturnResult(msgInner);
+            }
```
Comment (caigy) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:367]: If retry times of the half message reached maximum value, the broker where it escaped would not check it any more, and the broker it was born also had skipped it, that message would be lost.

Comment (caigy) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:367]: If retry times of the half message reached maximum value, the broker where it escaped would not check it any more, and the broker it was born also had skipped it, that message would be lost.

**Review by caigy — CHANGES_REQUESTED** [review]
I've left some comments in proposal, hope that it would be helpful.

**Comment by hzh0425** [comment]
Please add some tests to prove correctness

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java (hunk 3 above)
Comment (azhsmesos) [reply broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:367]: The transactionmessageserviceimpl#putbacktohalfqueuereturnresult() method needs to rewrite the half message. Here,I think we have to rewrite the topic. I understand that this step should also be carried out by escaping to other masters

**Comment by azhsmesos** [comment]
> Please add some tests to prove correctness

thanks！I will add it when i local test later

**Pull request closed without merging**